{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "686650a4-f255-4dd1-8851-c327baf22f30",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import multiprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "79e74878-cdc5-4bf8-815d-4715e547a945",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Количество процессоров: 6\n"
     ]
    }
   ],
   "source": [
    "def get_processor_count():\n",
    "    return multiprocessing.cpu_count()\n",
    "\n",
    "processor_count = get_processor_count()\n",
    "print(\"Количество процессоров:\", processor_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f236cf0-ae14-4eb4-856c-c476c64c17c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ[\"GATK_LOCAL_JAR\"] = \"./utils/gatk/gatk-package-4.3.0.0-local.jar\"\n",
    "os.makedirs(\"./results\", exist_ok=True)\n",
    "input_dir = \"./input\"\n",
    "log_dir = \"./logs\"\n",
    "os.makedirs(log_dir, exist_ok=True)\n",
    "\n",
    "panel = \"./bed/hg38_solotest.interval_list\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93369cd6-5588-4b09-a156-d535de1990a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.listdir(input_dir):\n",
    "    print(\"Empty input directory\")\n",
    "    exit(1)\n",
    "else:\n",
    "    names = []\n",
    "    for infile in os.listdir(input_dir):\n",
    "        if infile.endswith(\"_L00?_R1.fq.gz\"):\n",
    "            file = os.path.join(input_dir, infile)\n",
    "            base = os.path.splitext(infile)[0]\n",
    "            name = base.split(\"_\")[0]\n",
    "            if name not in names:\n",
    "                names.append(name)\n",
    "    with open(os.path.join(log_dir, base + \".log\"), \"w\") as logfile:\n",
    "        logfile.write(\"names=\" + repr(names) + \"\\n\")\n",
    "    for name in names:\n",
    "        files = []\n",
    "        lines = []\n",
    "        bams = []\n",
    "        with open(os.path.join(log_dir, base + \".log\"), \"a\") as logfile:\n",
    "            logfile.write(\"Processing sample: \" + name + \"\\n\")\n",
    "            files = [os.path.join(input_dir, f\"{name}_L00{line}_R1.fq.gz\") for line in lines]\n",
    "            logfile.write(\"files=\" + repr(files) + \"\\n\")\n",
    "            for infile in files:\n",
    "                base = os.path.splitext(os.path.basename(infile))[0]\n",
    "                line = base.split(\"_\")[1]\n",
    "                R1 = f\"{name}_{line}_R1.fq.gz\"\n",
    "                R2 = f\"{name}_{line}_R2.fq.gz\"\n",
    "                with open(os.path.join(log_dir, base + \".log\"), \"a\") as logfile:\n",
    "                    logfile.write(\"Processing line: \" + line + \"\\n\")\n",
    "                    logfile.write(\"Processing pair: \" + fR1 + \" \" + fR2 + \"\\n\")\n",
    "                    logfile.write(\"Adapter trimming...\\n\")\n",
    "                trimmomatic_dir = \"./results/trimmomatic\"\n",
    "                os.makedirs(trimmomatic_dir, exist_ok=True)\n",
    "                command = [\n",
    "                    \"java\",\n",
    "                    \"-jar\",\n",
    "                    \"./utils/Trimmomatic/trimmomatic-0.39.jar\",\n",
    "                    \"PE\",\n",
    "                    \"-threads\",\n",
    "                    str(processor_count),\n",
    "                    \"-phred33\",\n",
    "                    os.path.join(input_dir, R1),\n",
    "                    os.path.join(input_dir, R2),\n",
    "                    os.path.join(trimmomatic_dir, base + \"_R1.trim.fastq.gz\"),\n",
    "                    os.path.join(trimmomatic_dir, base + \"_R1un.trim.fastq.gz\"),\n",
    "                    os.path.join(trimmomatic_dir, base + \"_R2.trim.fastq.gz\"),\n",
    "                    os.path.join(trimmomatic_dir, base + \"_R2un.trim.fastq.gz\"),\n",
    "                    \"ILLUMINACLIP:./utils/Trimmomatic/TruSeq3-PE-2.fa:2:40:15:2:true\",\n",
    "                    \"LEADING:3\",\n",
    "                    \"TRAILING:3\",\n",
    "                    \"SLIDINGWINDOW:6:15\",\n",
    "                    \"MINLEN:36\",\n",
    "                ]\n",
    "                with open(os.path.join(log_dir, base + \".log\"), \"a\") as logfile:\n",
    "                    logfile.write(\"Converting to uBAM...\\n\")\n",
    "                subprocess.run(command, stdout=logfile, stderr=subprocess.STDOUT, text=True)\n",
    "                \n",
    "                with open(os.path.join(log_dir, base + \".log\"), \"a\") as logfile:\n",
    "                    logfile.write(\"Aligning...\\n\")\n",
    "                    samtofastq_command = [\n",
    "                        \"./utils/gatk/gatk\",\n",
    "                        \"--java-options\",\n",
    "                        \"-XX:-PrintFlagsFinal -XX:+UseG1GC -Dsamjdk.use_async_io_read_samtools=true -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -XX:GCTimeLimit=50 -XX:GCHeapFreeLimit=10 -Xms1000m -Xmx16000m\",\n",
    "                        \"SamToFastq\",\n",
    "                        \"--INPUT\",\n",
    "                        os.path.join(\"./results/FastqToSam\", base + \".unmapped.qsorted.bam\"),\n",
    "                        \"--FASTQ\",\n",
    "                        \"/dev/stdout\",\n",
    "                        \"--INTERLEAVE\",\n",
    "                        \"true\",\n",
    "                        \"--NON_PF\",\n",
    "                        \"true\",\n",
    "                    ]\n",
    "                    bwa_command = [\n",
    "                        \"bwa\",\n",
    "                        \"mem\",\n",
    "                        \"-K\",\n",
    "                        \"100000000\",\n",
    "                        \"-c\",\n",
    "                        \"250\",\n",
    "                        \"-p\",\n",
    "                        \"-v\",\n",
    "                        \"3\",\n",
    "                        \"-t\",\n",
    "                        str(processor_count),\n",
    "                        \"-Y\",\n",
    "                        \"./utils/bwa/GRCh38.fa\",\n",
    "                        \"/dev/stdin\",\n",
    "                        \"-\",\n",
    "                    ]\n",
    "                    mergebamalignment_command = [\n",
    "                        \"./utils/gatk/gatk\",\n",
    "                        \"--java-options\",\n",
    "                        \"-XX:-PrintFlagsFinal -XX:+UseG1GC -Dsamjdk.compression_level=5 -Dsamjdk.use_async_io_read_samtools=true -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -XX:GCTimeLimit=50 -XX:GCHeapFreeLimit=10 -Xms1000m -Xmx16000m\",\n",
    "                        \"MergeBamAlignment\",\n",
    "                        \"--VALIDATION_STRINGENCY\",\n",
    "                        \"SILENT\",\n",
    "                        \"--EXPECTED_ORIENTATIONS\",\n",
    "                        \"FR\",\n",
    "                        \"--ATTRIBUTES_TO_RETAIN\",\n",
    "                        \"X0\",\n",
    "                        \"--ATTRIBUTES_TO_REMOVE\",\n",
    "                        \"NM\",\n",
    "                        \"--ATTRIBUTES_TO_REMOVE\",\n",
    "                        \"MD\",\n",
    "                        \"--ALIGNED_BAM\",\n",
    "                        \"/dev/stdin\",\n",
    "                        \"--UNMAPPED_BAM\",\n",
    "                        os.path.join(\"./results/FastqToSam\", base + \".unmapped.qsorted.bam\"),\n",
    "                        \"--OUTPUT\",\n",
    "                        os.path.join(\"./results/BWA\", base + \".aligned.sorted.bam\"),\n",
    "                        \"--REFERENCE_SEQUENCE\",\n",
    "                        \"./utils/ref/GRCh38.fa\",\n",
    "                        \"--PAIRED_RUN\",\n",
    "                        \"true\",\n",
    "                        \"--SORT_ORDER\",\n",
    "                        \"coordinate\",\n",
    "                        \"--IS_BISULFITE_SEQUENCE\",\n",
    "                        \"false\",\n",
    "                        \"--ALIGNED_READS_ONLY\",\n",
    "                        \"false\",\n",
    "                        \"--CLIP_ADAPTERS\",\n",
    "                        \"false\",\n",
    "                        \"--MAX_RECORDS_IN_RAM\",\n",
    "                        \"5000000\",\n",
    "                        \"--ADD_MATE_CIGAR\",\n",
    "                        \"true\",\n",
    "                        \"--MAX_INSERTIONS_OR_DELETIONS\",\n",
    "                        \"-1\",\n",
    "                        \"--PRIMARY_ALIGNMENT_STRATEGY\",\n",
    "                        \"MostDistant\",\n",
    "                        \"--PROGRAM_RECORD_ID\",\n",
    "                        \"bwamem\",\n",
    "                        \"--PROGRAM_GROUP_VERSION\",\n",
    "                        \"0.7.17\",\n",
    "                        \"--PROGRAM_GROUP_COMMAND_LINE\",\n",
    "                        \"bwa mem -K 100000000 -M -c 250 -p -v 3 -t \" + str(processor_count) + \" -Y ./utils/bwa/GRCh38.fa\",\n",
    "                        \"--PROGRAM_GROUP_NAME\",\n",
    "                        \"bwamem\",\n",
    "                        \"--UNMAPPED_READ_STRATEGY\",\n",
    "                        \"COPY_TO_TAG\",\n",
    "                        \"--ALIGNER_PROPER_PAIR_FLAGS\",\n",
    "                        \"true\",\n",
    "                        \"--UNMAP_CONTAMINANT_READS\",\n",
    "                        \"true\",\n",
    "                        \"--CREATE_INDEX\",\n",
    "                        \"true\",\n",
    "                        \"--ADD_PG_TAG_TO_READS\",\n",
    "                        \"false\",\n",
    "                    ]\n",
    "                    \n",
    "                subprocess.run(samtofastq_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "                samtofastq_output = subprocess.Popen(\n",
    "                    bwa_command,\n",
    "                    stdin=subprocess.PIPE,\n",
    "                    stdout=subprocess.PIPE,\n",
    "                    stderr=subprocess.PIPE,\n",
    "                    text=True,\n",
    "                )\n",
    "                mergebamalignment_output = subprocess.run(\n",
    "                    mergebamalignment_command,\n",
    "                    stdin=samtofastq_output.stdout,\n",
    "                    stdout=subprocess.PIPE,\n",
    "                    stderr=subprocess.PIPE,\n",
    "                    text=True,\n",
    "                )\n",
    "                samtofastq_output.stdout.close()\n",
    "                    \n",
    "                bams += \"-I ./results/BWA/\" + base + \".aligned.sorted.bam \"\n",
    "                    \n",
    "                os.makedirs(os.path.join(results_dir, \"FastqToSam\"), exist_ok=True)\n",
    "                os.makedirs(os.path.join(results_dir, \"BQSR\"), exist_ok=True)\n",
    "                os.makedirs(os.path.join(results_dir, \"GATK\"), exist_ok=True)\n",
    "                ampliconmetrics_dir = \"./ampliconmetrics\"\n",
    "                os.makedirs(ampliconmetrics_dir, exist_ok=True)\n",
    "                cover_dir = \"./cover\"\n",
    "                os.makedirs(cover_dir, exist_ok=True)\n",
    "                    \n",
    "                with open(os.path.join(log_dir, base + \".log\"), \"a\") as logfile:\n",
    "                    logfile.write(\"Variants recalibrations...\\n\")\n",
    "                    bqsrt_command = [\n",
    "                        \"./utils/gatk/gatk\",\n",
    "                        \"--java-options\",\n",
    "                        \"-XX:-PrintFlagsFinal -XX:+UseParallelGC -Dsamjdk.compression_level=5 -XX:GCTimeLimit=50 -XX:GCHeapFreeLimit=10 -Xms1000m -Xmx8000m\",\n",
    "                        \"BaseRecalibrator\",\n",
    "                        \"-R\",\n",
    "                        \"./utils/ref/GRCh38.fa\",\n",
    "                        bams,\n",
    "                        \"--use-original-qualities\",\n",
    "                        \"true\",\n",
    "                        \"-O\",\n",
    "                        os.path.join(results_dir, \"BQSR\", name + \".stats\"),\n",
    "                        \"--known-sites\",\n",
    "                        \"./utils/ref/dbsnp_138.hg38.vcf.gz\",\n",
    "                        \"--known-sites\",\n",
    "                        \"./utils/ref/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz\",\n",
    "                        \"-L\",\n",
    "                        panel,\n",
    "                    ]\n",
    "                    applybqsr_command = [\n",
    "                        \"./utils/gatk/gatk\",\n",
    "                        \"--java-options\",\n",
    "                        \"-XX:-PrintFlagsFinal -XX:+UseParallelGC -Dsamjdk.compression_level=5 -XX:GCTimeLimit=50 -XX:GCHeapFreeLimit=10 -Xms1000m -Xmx8000m\",\n",
    "                        \"ApplyBQSR\",\n",
    "                        \"--create-output-bam-md5\",\n",
    "                        \"--add-output-sam-program-record\",\n",
    "                        \"true\",\n",
    "                        \"-R\",\n",
    "                        \"./utils/ref/GRCh38.fa\",\n",
    "                        bams,\n",
    "                        \"--use-original-qualities\",\n",
    "                        \"true\",\n",
    "                        \"-O\",\n",
    "                        os.path.join(results_dir, \"BQSR\", name + \".recalibrated.aligned.bam\"),\n",
    "                        \"--create-output-bam-index\",\n",
    "                        \"--bqsr-recal-file\",\n",
    "                        os.path.join(results_dir, \"BQSR\", name + \".stats\"),\n",
    "                        \"--static-quantized-quals\",\n",
    "                        \"10\",\n",
    "                        \"--static-quantized-quals\",\n",
    "                        \"20\",\n",
    "                        \"--static-quantized-quals\",\n",
    "                        \"30\",\n",
    "                        \"-L\",\n",
    "                        panel,\n",
    "                    ]\n",
    "                with open(os.path.join(log_dir, base + \".log\"), \"a\") as logfile:\n",
    "                    logfile.write(\"Variant calling...\\n\")\n",
    "                    haplotypecaller_command = [\n",
    "                        \"./utils/gatk/gatk\",\n",
    "                        \"--java-options\",\n",
    "                        \"-XX:-PrintFlagsFinal -XX:+UseParallelGC -XX:ParallelGCThreads=2 -Dsamjdk.use_async_io_read_samtools=true -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -XX:GCTimeLimit=50 -XX:GCHeapFreeLimit=10 -Xms1000m -Xmx8000m\",\n",
    "                        \"HaplotypeCaller\",\n",
    "                        \"-R\",\n",
    "                        \"./utils/ref/GRCh38.fa\",\n",
    "                        \"-I\",\n",
    "                        os.path.join(results_dir, \"BQSR\", name + \".recalibrated.aligned.bam\"),\n",
    "                        \"-L\",\n",
    "                        panel,\n",
    "                        \"-O\",\n",
    "                        os.path.join(results_dir, \"GATK\", name + \".vcf.gz\"),\n",
    "                        \"--max-alternate-alleles\",\n",
    "                        \"6\",\n",
    "                        \"--max-reads-per-alignment-start\",\n",
    "                        \"0\",\n",
    "                        \"--pair-hmm-implementation\",\n",
    "                        \"FASTEST_AVAILABLE\",\n",
    "                        \"--interval-padding\",\n",
    "                        \"25\",\n",
    "                        \"--native-pair-hmm-threads\",\n",
    "                        str(processor_count),\n",
    "                        \"--smith-waterman\",\n",
    "                        \"JAVA\",\n",
    "                        \"-A\",\n",
    "                        \"Coverage\",\n",
    "                        \"-A\",\n",
    "                        \"ChromosomeCounts\",\n",
    "                        \"-A\",\n",
    "                        \"BaseQuality\",\n",
    "                        \"-A\",\n",
    "                        \"FragmentLength\",\n",
    "                        \"-A\",\n",
    "                        \"BaseQualityRankSumTest\",\n",
    "                        \"-A\",\n",
    "                        \"MappingQuality\",\n",
    "                        \"-A\",\n",
    "                        \"ReadPosRankSumTest\",\n",
    "                        \"-A\",\n",
    "                        \"FisherStrand\",\n",
    "                        \"-A\",\n",
    "                        \"ClippingRankSumTest\",\n",
    "                        \"-A\",\n",
    "                        \"StrandOddsRatio\",\n",
    "                        \"-A\",\n",
    "                        \"ReadPosition\",\n",
    "                    ]\n",
    "        subprocess.run(bqsrt_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)\n",
    "        subprocess.run(applybqsr_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)\n",
    "        subprocess.run(haplotypecaller_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
